lib/gds_api/test_helpers/email_alert_api.rb

```diff
@@ -68,6 +68,54 @@ def stub_email_alert_api_has_subscription(
           .to_return(status: 200, body: response)
       end
 
+      # Stubs the API responses as if each subscription happened in the order they are passed.
+      # Useful if you need to query the '/latest' endpoint of a subscription.
+      # Takes an array of hashes.
+      #
+      # @example
+      #  stub_email_alert_api_has_subscriptions([
+      #    {
+      #      id: 'id-of-my-subscriber-list',
+      #      frequency: 'weekly',
+      #      ended: true,
+      #    },
+      #    {
+      #      id: 'id-of-my-subscriber-list',
+      #      frequency: 'daily',
+      #    },
+      #  ])
+      #
+      # @param subscriptions [Array]
+      def stub_email_alert_api_has_subscriptions(subscriptions)
+        subscriptions.map! { |subscription| apply_subscription_defaults(subscription) }
+        subscriptions.each do |id, params|
```

Comment:
Are two loops needed here?

Reply:
Yes, as we pass the modified `subscriptions` as a parameter on the next line. We've got a nice API for calling the method, and we need to transform the data into something we can pass around.

Having two loops as we do here is a complexity of `2n`; the alternative is a loop within a loop (`n²`).

Reply:
Would it be possible to have something like this to avoid two loops?

```
subscriptions.each do |subscription|
  subscription = apply_subscription_defaults(subscription)
  ...
```

Reply:
Not reeeeeally, because we'd be using `subscriptions` before all of the subscriptions have been transformed, like:

```ruby
subscriptions.each do |subscription|
  subscription = apply_subscription_defaults(subscription)
  foo(subscriptions) # oops - we now need to apply defaults to all of these somehow
```

Which is why I've done it outside the loop:

```ruby
subscriptions = transform_the_subs(subscriptions)
subscriptions.each do |subscription|
  subscription = apply_subscription_defaults(subscription)
  foo(subscriptions) # now we can call this safely :)
```

If it's code _efficiency_ we're worried about, I think it's better as-is, otherwise within `foo(subscriptions)` we have to call `apply_subscription_defaults` on each subscription (a loop within a loop).

If it's just readability we're worried about, I can look at calling `apply_subscription_defaults` in the `get_latest_matching` method...?

Reply:
Ah, yes, I see now.  I cannot see any way the two loops could be condensed into one without any further code complexity elsewhere.